File: 0785-is-graph-bipartite/0785-is-graph-bipartite.cpp
class Solution
{
    bool contradiction{false};

public:
    void DFS(vector<vector<int>> &graph, vector<int> &colors, int node, int colorToAdd = 1)
    {
        if (colors[node] == 0)
            colors[node] = colorToAdd;
        else
        {
            if (colors[node] != colorToAdd)
            {
                contradiction = true;
            }

            return;
        }
        for (auto neighbour : graph[node])
        {
            DFS(graph, colors, neighbour, 3 - colorToAdd); // 3-1 = 2 ;   3-2 = 1 ;
        }
    }

    bool isBipartite(vector<vector<int>> &graph)
    {
        vector<int> colors(graph.size());

        // for CC
        for (int i = 0; i < graph.size(); i++)
        {
            if (colors[i] == 0)
                DFS(graph, colors, i);
            if (contradiction)
                return false;
        }

        return true;
    }
};

File: 0785-is-graph-bipartite/0785-is-graph-bipartite.java
import java.util.Arrays;
import java.util.LinkedList;
import java.util.Queue;

class Solution {


    /*
     * Two colors problem
     *
     * */
    public boolean isBipartite(int[][] graph) {
        int[] visited = new int[graph.length];
        for (int i = 0; i < graph.length; i++) {
            visited[i]=0;
        }
        for (int i = 0; i < graph.length; i++) {
            if (visited[i] != 0) continue;
            Queue<Integer> queue = new LinkedList<>();
            visited[i] = 1;
            queue.add(i);
            while (!queue.isEmpty()) {
                int node = queue.poll();
                for (int neighbour : graph[node]) {
                    if (visited[neighbour] == visited[node]) {
                        return false;
                    } else if (visited[neighbour] == 0) {
                        visited[neighbour] = visited[node] * -1;
                        queue.add(neighbour);
                    }
                }


            }
        }

        return true;
    }
}